File: flutter_app/lib/models/audit_log.dart
class AuditLog {
  final int? logId;
  final String? tableName;
  final String? operation;
  final String? userId;
  final String? username;
  final String? recordId;
  final String? oldValues;
  final String? newValues;
  final DateTime? operationDate;
  final String? ipAddress;
  final String? sessionId;

  AuditLog({
    this.logId,
    this.tableName,
    this.operation,
    this.userId,
    this.username,
    this.recordId,
    this.oldValues,
    this.newValues,
    this.operationDate,
    this.ipAddress,
    this.sessionId,
  });

  factory AuditLog.fromJson(Map<String, dynamic> json) {
    return AuditLog(
      logId: json['logId'] as int?,
      tableName: json['tableName'] as String?,
      operation: json['operation'] as String?,
      userId: json['userId'] as String?,
      username: json['username'] as String?,
      recordId: json['recordId'] as String?,
      oldValues: json['oldValues'] as String?,
      newValues: json['newValues'] as String?,
      operationDate: json['operationDate'] != null
          ? DateTime.tryParse(json['operationDate'])
          : null,
      ipAddress: json['ipAddress'] as String?,
      sessionId: json['sessionId'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'logId': logId,
      'tableName': tableName,
      'operation': operation,
      'userId': userId,
      'username': username,
      'recordId': recordId,
      'oldValues': oldValues,
      'newValues': newValues,
      'operationDate': operationDate?.toIso8601String(),
      'ipAddress': ipAddress,
      'sessionId': sessionId,
    };
  }

  bool get isInsert => operation == 'INSERT';
  bool get isUpdate => operation == 'UPDATE';
  bool get isDelete => operation == 'DELETE';
  bool get isSelect => operation == 'SELECT';

  String get operationDisplay {
    switch (operation) {
      case 'INSERT':
        return 'Created';
      case 'UPDATE':
        return 'Updated';
      case 'DELETE':
        return 'Deleted';
      case 'SELECT':
        return 'Viewed';
      default:
        return operation ?? 'Unknown';
    }
  }
}

File: backend/src/main/java/edu/university/grademanagement/model/entity/AuditLog.java
package edu.university.grademanagement.model.entity;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;

/**
 * AuditLog Entity
 * Represents audit trail entries for data changes
 */
@Entity
@Table(name = "AUDIT_LOG", schema = "GMS_ADMIN")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class AuditLog {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "log_id")
    private Long logId;

    @Column(name = "table_name", nullable = false, length = 30)
    private String tableName;

    @Column(name = "operation", nullable = false, length = 10)
    private String operation;

    @Column(name = "user_id", length = 10)
    private String userId;

    @Column(name = "username", length = 50)
    private String username;

    @Column(name = "record_id", length = 50)
    private String recordId;

    @Lob
    @Column(name = "old_values")
    private String oldValues;

    @Lob
    @Column(name = "new_values")
    private String newValues;

    @Column(name = "operation_date")
    private LocalDateTime operationDate;

    @Column(name = "ip_address", length = 50)
    private String ipAddress;

    @Column(name = "session_id", length = 100)
    private String sessionId;

    @PrePersist
    protected void onCreate() {
        if (operationDate == null) {
            operationDate = LocalDateTime.now();
        }
    }

    public boolean isInsert() {
        return "INSERT".equals(operation);
    }

    public boolean isUpdate() {
        return "UPDATE".equals(operation);
    }

    public boolean isDelete() {
        return "DELETE".equals(operation);
    }

    public boolean isSelect() {
        return "SELECT".equals(operation);
    }
}

